extensions/smallrye-fault-tolerance/runtime/src/main/java/io/quarkus/smallrye/faulttolerance/runtime/QuarkusFaultToleranceOperationProvider.java

```diff
@@ -1,22 +1,50 @@
 package io.quarkus.smallrye.faulttolerance.runtime;
 
 import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.function.Function;
 
 import javax.annotation.Priority;
-import javax.enterprise.context.Dependent;
 import javax.enterprise.inject.Alternative;
+import javax.inject.Singleton;
+
+import org.jboss.logging.Logger;
 
 import io.smallrye.faulttolerance.FaultToleranceOperationProvider;
 import io.smallrye.faulttolerance.config.FaultToleranceOperation;
 
-@Dependent
+@Singleton
 @Alternative
 @Priority(1)
 public class QuarkusFaultToleranceOperationProvider implements FaultToleranceOperationProvider {
+    private static final Logger LOG = Logger.getLogger(QuarkusFaultToleranceOperationProvider.class);
+    private Map<String, FaultToleranceOperation> operationCache;
+
+    /**
+     * Called by SmallryeFaultToleranceRecorder to init the operation cache.
+     */
+    void init(Map<String, FaultToleranceOperation> operationCache) {
+        this.operationCache = operationCache;
+    }
 
     @Override
     public FaultToleranceOperation get(Class<?> beanClass, Method method) {
+        String key = getCacheKey(beanClass, method);
+        return operationCache.computeIfAbsent(key, new Function<String, FaultToleranceOperation>() {
```

Comment:
I don't see why we don't make the function a constant? Is there something forbidding it?

Reply:
Because we use the method arguments inside the function -> `createAtRuntime(beanClass, method, key)`.

Reply:
It's possible if I used a CompositeKey with beanClass and method instead of a String. So everything needed to create the operation will be on the key and the Function could be a constant.

Reply:
Yeah, I was thinking about that.

TBH, the best solution here is to do what I suggested first: we do a first `get()` call and only call `computeIfAbsent` if we have nothing. As previously mentioned, I did some extensive benchmarks on HV and it's significantly faster if this is called often (and I suppose it will be called for every call to a function that is marked with FT annotations or am I missing something?). And this will avoid creating the function if this is in the cache.

Reply:
That would not help much - you'll have to create a new key object for each `get()` anyway...

Reply:
Previsouly I created a new String for each object so it should be the same ...

Reply:
And the key object just hold two references for existing object so should be small regarding a new string ...

Reply:
@gsmet @mkouba updated with a composite CacheKey to avoid re-creating the Function on each call and implemented @gsmet optimization on the CHM.

If it's OK I'll squash the last commit and we'll be ready to go :)

Reply:
@mkouba you won't avoid creating the key but you can avoid creating the function.